### DISCUSSION TITLE
Connection management

### DISCUSSION THREAD

**Discussion by tthebst** [libp2p/rust-libp2p#3670] [Q&A] [answered]
Hey!

I'm trying to better understand the concept of connections in libp2p. Few questions:

- Libp2p allows having many connections to the same peer, but what is the benefit of having multiple connections? Isn't one connection enough if your application runs on a single port and there only is one multiaddr for one peer? 
- Does it makes sense to set [with_max_established_per_peer](https://docs.rs/libp2p/latest/libp2p/swarm/struct.ConnectionLimits.html#method.with_max_established_per_peer) to one?
- If you have multiple connections open to peer and you send a message (for example req/resp) how does libp2p decide which connection to use? 
- Is there one connection handler task per connection per behavior or one connection handler per connection?
- With recent changes, it seems possible to write a "Peer Manager" network behaviour that does connection management (and not go through the swarm). Is this recommended?

**Comment by thomaseizinger** [comment, marked as answer]
> Hey!
> 
> I'm trying to better understand the concept of connections in libp2p. Few questions:

All excellent questions!

>     * Libp2p allows having many connections to the same peer, but what is the benefit of having multiple connections? Isn't one connection enough if your application runs on a single port and there only is one multiaddr for one peer?

Technically, it is enough. Practically, you can't avoid it. How would you decide which connection to keep if two peers dial each other at the same time?

>     * Does it makes sense to set [with_max_established_per_peer](https://docs.rs/libp2p/latest/libp2p/swarm/struct.ConnectionLimits.html#method.with_max_established_per_peer) to one?

This function is deprecated in master already, see https://github.com/libp2p/rust-libp2p/pull/3386. You can set it to 1 but may experience weird behaviour because both peers may close a different connection and you end up with none.

>     * If you have multiple connections open to peer and you send a message (for example req/resp) how does libp2p decide which connection to use?

A `NetworkBehaviour` can decide which connection to send a message to: https://docs.rs/libp2p/latest/libp2p/swarm/enum.NotifyHandler.html

The request-respones behaviour in particular uses a round-robin approach across all active connections: https://github.com/libp2p/rust-libp2p/blob/3f99c31bb368729a05159133cf2bf4aef8081aa9/protocols/request-response/src/lib.rs#L538

>     * Is there one connection handler task per connection per behavior or one connection handler per connection?

Behaviours are organised as "slices", i.e. they all share all connections. Each connection has a connection handler which is again composed out of all the handlers of all behaviours. It is one task per connection but the handlers are composed.

>     * With recent changes, it seems possible to write a "Peer Manager" network behaviour that does connection management (and not go through the swarm). Is this recommended?

Yes, see:

- https://github.com/libp2p/rust-libp2p/pull/3386
- https://github.com/libp2p/rust-libp2p/pull/3590

**Comment by tthebst** [comment]
Thanks for the response!

So is there any benefit of having more than one or two open connections?

**Reply by mxinden** [reply]
In general, there is no benefit of having more than one connection between the same set of two peers.

On TCP one could argue that multiple connections prevent head-of-line blocking between two concurrent streams. Though I would consider that a potential optimization worth measuring first.

QUIC does not have head-of-line blocking on the connection level, thus I would not expect there to be any gains from having more than one connection between two peers.

Happy to go into detail @tthebst. Let me know if this is helpful.

**Reply by rkuhn** [reply]
There is a benefit to allowing a second connection: how else could the dialled peer validate that it can connect back to the dialling peer? Curating known-good addresses is a difficult and important part of p2p systems.

**Reply by mxinden** [reply]
True. My comment above only addressed "So is there any benefit of having more than one or two [**long lived**] open connections?"

Agreed that there is value in testing both directions. The second connection (in the reverse direction) can be short lived.
